verifiers/proofofsql/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+pub struct ProofOfSql<T>;
+
+impl<T: Config> Verifier for ProofOfSql<T> {
+    type Proof = Vec<u8>;
+
+    type Pubs = Vec<u8>;
+
+    type Vk = Vk<T>;
+
+    fn hash_context_data() -> &'static [u8] {
+        b"proofofsql-v0.28"
+    }
+
+    fn verify_proof(
+        vk: &Self::Vk,
+        proof: &Self::Proof,
+        pubs: &Self::Pubs,
+    ) -> Result<(), VerifyError> {
+        vk.validate_size()?;
+        let proof = proof_of_sql_verifier::Proof::try_from(&proof[..])
+            .map_err(Into::<ErrorWrapper>::into)?;
+        let pubs = proof_of_sql_verifier::PublicInput::try_from(&pubs[..])
+            .map_err(Into::<ErrorWrapper>::into)?;
+        let vk = proof_of_sql_verifier::VerificationKey::try_from(&vk.0[..])
+            .map_err(Into::<ErrorWrapper>::into)?;
+        proof_of_sql_verifier::verify_proof(&proof, &pubs, &vk)
+            .map_err(Into::<ErrorWrapper>::into)?;
+        Ok(())
+    }
+
+    fn validate_vk(vk: &Self::Vk) -> Result<(), VerifyError> {
+        vk.validate_size()
+            .inspect_err(|_| log::debug!("Verification key is too big"))?;
+        VerificationKey::try_from(vk.0.as_slice())
+            .inspect_err(|e| log::debug!("Cannot parse verification key: {:?}", e))
+            .map_err(|_| VerifyError::InvalidVerificationKey)?;
+        Ok(())
+    }
+
+    fn pubs_bytes(pubs: &Self::Pubs) -> Cow<[u8]> {
+        let data = pubs.to_vec();
+        Cow::Owned(data)
+    }
+}
+
+pub struct ProofOfSqlWeight<W: weight::WeightInfo>(PhantomData<W>);
+
+impl<T: Config, W: weight::WeightInfo> pallet_verifiers::WeightInfo<ProofOfSql<T>>
+    for ProofOfSqlWeight<W>
+{
+    fn submit_proof(
+        _proof: &<ProofOfSql<T> as hp_verifiers::Verifier>::Proof,
+        _pubs: &<ProofOfSql<T> as hp_verifiers::Verifier>::Pubs,
+    ) -> Weight {
+        W::submit_proof()
+    }
+
+    fn submit_proof_with_vk_hash(
+        _proof: &<ProofOfSql<T> as hp_verifiers::Verifier>::Proof,
+        _pubs: &<ProofOfSql<T> as hp_verifiers::Verifier>::Pubs,
+    ) -> Weight {
+        W::submit_proof_with_vk_hash()
+    }
+
+    fn register_vk(_vk: &<ProofOfSql<T> as hp_verifiers::Verifier>::Vk) -> Weight {
+        W::register_vk()
+    }
+}
+
+#[cfg(test)]
+mod vk {
+    use super::*;
+
+    use frame_support::{assert_err, assert_ok};
+    use sp_core::ConstU32;
+
+    pub struct ConfigWithMaxNuEqualTo5;
+
+    impl Config for ConfigWithMaxNuEqualTo5 {
+        type LargestMaxNu = ConstU32<5>;
+    }
+
+    pub struct ConfigWithMaxNuEqualTo4;
+
+    impl Config for ConfigWithMaxNuEqualTo4 {
+        type LargestMaxNu = ConstU32<4>;
+    }
+
+    pub struct ConfigWithMaxNuEqualTo3;
+
+    impl Config for ConfigWithMaxNuEqualTo3 {
+        type LargestMaxNu = ConstU32<3>;
+    }
+
+    #[test]
+    fn length_should_match_max_encoded_len() {
+        let vk_with_nu_equal_to_4: <ProofOfSql<ConfigWithMaxNuEqualTo4> as Verifier>::Vk =
+            include_bytes!("resources/VALID_VK.bin").to_vec().into();
```

Comment:
Just to remove this duplication: can you provide a function to get the bytes vector?

Reply:
Done. Also took the chance to refactor the tests a little bit in https://github.com/HorizenLabs/zkVerify/pull/136/commits/7cefb1ee705073d20c636cfd45c4799b2f56ea3e